# Mini Project 1 — Analysis Notebook

**Your name: Michelle Tran**

**Dataset: Patient Satisfaction**

**Date: 5/10/26**

This notebook has four sections. Work through them in order. Each section has instructions and a code cell to fill in. Add markdown cells to explain your thinking as you go — that writing is part of the assignment.

When you're done, publish this notebook to your GitHub repository and submit the URL to Canvas.

In [2]:
# Setup — run this cell first
# If any package is missing, it will install automatically
import subprocess, sys

def install(pkg):
    subprocess.check_call([sys.executable, "-m", "pip", "install", pkg, "-q"])

for pkg in ["pandas", "plotly", "kaleido", "nbformat"]:
    try:
        __import__(pkg)
    except ImportError:
        print(f"Installing {pkg}...")
        install(pkg)

import pandas as pd
import plotly.express as px

print("Setup complete.")

Setup complete.


---

## Section 1 — Overview

Before writing any code, fill in this section. A good Overview tells anyone reading your notebook — including a future employer — what the analysis is about before they see a single chart.

**Dataset:** *Patient Satisfaction Dataset https://www.kaggle.com/datasets/vdimitrievska/patient-satisfaction-dataset?resource=download*

**I want to explore the factors that affect patient satisfaction in health care services. This is important to me because there should be a more functional system that welcomes patients and considers the diversity of patients' needs.*

**Three analytical questions:**

1. *Which services are most strongly associated with overall satisfaction?*
2. *Do personable interactions with healthcare workers, such as doctor communication and staff friendliness, show a stronger association with overall patient satisfaction than operational factors, such as wait times, administrative procedures, and facility quality?*
3. *Do patients who rate clinical quality highly also tend to rate operational factors highly?*

**What a practitioner would do with these findings:** *These findings would help a practitioner understand patient priorities and what is associated with their dissatisfaction in healthcare services.*

---

## Section 2 — Data Profile

Load your dataset and get a basic picture of what's in it. Answer these questions in a markdown cell below your code:

- How many rows and columns does your dataset have?
- What does each column represent?
- Are there any obvious data quality issues (missing values, unexpected types, inconsistent formatting)?
- Which column or columns will your analysis focus on, and why?

In [ ]:
df = pd.read_csv('patient_satisfaction_dataset.csv')  # ← replace with your filename
#The original ratings (1:satisfied, 2: no satisfied, 3:neutral, 4: very satisfied, 5:very unsatisfied) were reordered to (1: very unsatisfied, 2: not satisfied, 3:neutral, 4: satisfied, 5:very satisfied) to make it easier to analyze the data.
RECODE_TO_ORDERED = { # <-- Recoding dictionary
    5: 1,  # very unsatisfied
    2: 2,  # not satisfied
    3: 3,  # neutral
    1: 4,  # satisfied
    4: 5,  # very satisfied
}

for col in df.columns:
    df[col] = df[col].map(RECODE_TO_ORDERED) # .map is used to apply the recoding dictionary I created to each element in the column

# Based on the null results from df.isnull().sum(), I used the dropna() function to remove all rows with missing values.
# Original rows: 453
df = df.dropna()

print(df.shape)
df.head(500)

(452, 17)


,satisfaction in RM,Check up appointment,Time waiting,Admin procedures,Hygiene and cleaning,Time of appointment,Quality/experience dr.,Specialists avaliable,Communication with dr,Exact diagnosis,Modern equipment,friendly health care workers,lab services,avaliablity of drugs,waiting rooms,hospital rooms quality,"parking, playing rooms, caffes"
0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0
1,2.0,2.0,4.0,2.0,4.0,2.0,2.0,4.0,2.0,4.0,2.0,4.0,4.0,4.0,4.0,2.0,4.0
2,2.0,4.0,4.0,2.0,2.0,2.0,5.0,5.0,5.0,5.0,4.0,5.0,5.0,4.0,2.0,2.0,2.0
3,2.0,2.0,5.0,4.0,4.0,2.0,5.0,4.0,4.0,5.0,4.0,4.0,5.0,1.0,5.0,4.0,4.0
4,3.0,5.0,4.0,4.0,2.0,4.0,1.0,1.0,1.0,1.0,2.0,1.0,1.0,1.0,4.0,4.0,4.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
448,2.0,3.0,3.0,3.0,3.0,2.0,2.0,3.0,3.0,2.0,2.0,2.0,2.0,2.0,2.0,2.0,2.0
449,2.0,5.0,2.0,4.0,3.0,4.0,5.0,5.0,3.0,2.0,3.0,2.0,3.0,5.0,2.0,3.0,5.0
450,2.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0
451,3.0,4.0,3.0,3.0,3.0,3.0,3.0,2.0,4.0,3.0,2.0,3.0,3.0,3.0,3.0,3.0,3.0


In [7]:
# Check column types and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 452 entries, 0 to 452
Data columns (total 17 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   satisfaction in RM              452 non-null    float64
 1   Check up appointment            452 non-null    float64
 2   Time waiting                    452 non-null    float64
 3   Admin procedures                452 non-null    float64
 4   Hygiene and cleaning            452 non-null    float64
 5   Time of appointment             452 non-null    float64
 6   Quality/experience dr.          452 non-null    float64
 7   Specialists avaliable           452 non-null    float64
 8   Communication with dr           452 non-null    float64
 9   Exact diagnosis                 452 non-null    float64
 10  Modern equipment                452 non-null    float64
 11  friendly health care workers    452 non-null    float64
 12  lab services                    452 non-n

In [8]:
# Summary statistics for numeric columns
df.describe()

,satisfaction in RM,Check up appointment,Time waiting,Admin procedures,Hygiene and cleaning,Time of appointment,Quality/experience dr.,Specialists avaliable,Communication with dr,Exact diagnosis,Modern equipment,friendly health care workers,lab services,avaliablity of drugs,waiting rooms,hospital rooms quality,"parking, playing rooms, caffes"
count,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000,452.000000
mean,2.482301,3.353982,3.351770,3.234513,2.940265,3.278761,3.353982,3.190265,3.110619,3.398230,3.066372,3.117257,3.331858,3.212389,3.050885,3.006637,3.026549
std,0.530361,1.309494,1.281861,1.283653,1.220108,1.310774,1.460009,1.333216,1.400403,1.405026,1.229721,1.363758,1.387481,1.341484,1.213221,1.201422,1.127839
min,2.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
25%,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000
50%,2.000000,4.000000,4.000000,3.000000,3.000000,4.000000,4.000000,3.000000,3.000000,4.000000,3.000000,3.000000,4.000000,3.000000,3.000000,3.000000,3.000000
75%,3.000000,4.000000,4.000000,4.000000,4.000000,4.000000,5.000000,4.000000,4.000000,5.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000
max,4.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000


**Your data profile notes:**  
*Before starting, I noticed that the scale was rated out of order (1:satisfied, 2: no satisfied, 3:neutral, 4: very satisfied, 5:very unsatisfied). The original order was confusing so I made sure to recode it to the standard 1-5 Likert scale. The dataset originally showed 453 rows, but due to a row showing missing values I had added df = df.dropna() to filter out any row with missing values. Column 1 represents the overall satisfaction while columns 2-17 represent the satisfaction of different factors users experienced in the hospital. I'll be focusing on all columns to identify which out of the different factors has the highest satisfaction rating and what might be associated with the overall satisfaction.*

---

## Section 3 — Analysis

Answer your three research questions using pandas. Each question should have:

1. A markdown cell stating the question
2. A code cell with the analysis
3. A markdown cell with your interpretation — what does the result mean?

You may need to clean or reshape the data before you can answer a question. That's normal — document what you did and why.

**Question 1:** *Which services are most strongly associated with overall satisfaction?*

In [9]:
sat = "satisfaction in RM"
predictors = [c for c in df.columns if c != sat]

# Correlation of each aspect with overall satisfaction
assoc = (
    df[predictors]
    .corrwith(df[sat], method="spearman")
    .sort_values(ascending=False)
)

print("Strongest positive associations with overall satisfaction:")
print(assoc.head(10))
print("\nWeakest associations:")
print(assoc.tail(5))

# Optional: tidy table for your write-up
assoc_df = assoc.reset_index()
assoc_df.columns = ["aspect", "spearman_rho"]
assoc_df


Strongest positive associations with overall satisfaction:
Communication with dr           0.195961
Modern equipment                0.179516
Check up appointment            0.177963
friendly health care workers    0.172097
Hygiene and cleaning            0.157368
Specialists avaliable           0.150588
Exact diagnosis                 0.147300
Time waiting                    0.146718
Admin procedures                0.140284
Quality/experience dr.          0.122083
dtype: float64

Weakest associations:
Time of appointment               0.112084
avaliablity of drugs              0.100213
parking, playing rooms, caffes    0.099723
lab services                      0.099607
waiting rooms                     0.086547
dtype: float64


,aspect,spearman_rho
0,Communication with dr,0.195961
1,Modern equipment,0.179516
2,Check up appointment,0.177963
3,friendly health care workers,0.172097
4,Hygiene and cleaning,0.157368
5,Specialists avaliable,0.150588
6,Exact diagnosis,0.147300
7,Time waiting,0.146718
8,Admin procedures,0.140284
9,Quality/experience dr.,0.122083


**Interpretation:**  
*The factor that is the most strongly associated with the overall average is "Communication with dr" with 0.196 with "Modern equipment," "Check up appointment," "friendly health care workers", and "Hygiene and cleaning" following after.*

**Question 2:** *Do personable interactions with healthcare workers, such as doctor communication and staff friendliness, show a stronger association with overall patient satisfaction than operational factors, such as wait times, administrative procedures, and facility quality?*

In [12]:
sat = "satisfaction in RM"

personable = [
    "Communication with dr",
    "friendly health care workers",
    "Quality/experience dr.",
]

operational = [
    "Time waiting",
    "Admin procedures",
    "Hygiene and cleaning",
    "waiting rooms",
    "hospital rooms quality",
    "parking, playing rooms, caffes",
    "Check up appointment",
    "Time of appointment",
    "Modern equipment",
    "lab services",
    "avaliablity of drugs",
    "Specialists avaliable",
    "Exact diagnosis",
]
comp_personable = df[personable].mean(axis=1)
comp_operational = df[operational].mean(axis=1)

rho_personable = comp_personable.corr(df[sat], method="spearman")
rho_operational = comp_operational.corr(df[sat], method="spearman")

print(f"Personable composite vs {sat}: ρ = {rho_personable:.3f}")
print(f"Operational composite vs {sat}: ρ = {rho_operational:.3f}")

Personable composite vs satisfaction in RM: ρ = 0.207
Operational composite vs satisfaction in RM: ρ = 0.212


**Interpretation:**  
*The operational composite vs satisfaction in RM shows a stronger association (ρ = 0.212) compared to the personable composite vs satisfaction in RM (ρ = 0.212)*

**Question 3:** *Do patients who rate clinical quality highly also tend to rate operational factors highly?*

In [13]:
clinical = [
    "Quality/experience dr.",
    "Exact diagnosis",
    "Specialists avaliable",
    "Modern equipment",
    "lab services",
    "Communication with dr",
]

operational_q3 = [
    "Time waiting",
    "Admin procedures",
    "Hygiene and cleaning",
    "Check up appointment",
    "Time of appointment",
    "waiting rooms",
    "hospital rooms quality",
    "parking, playing rooms, caffes",
    "avaliablity of drugs",
]

clinical_avg = df[clinical].mean(axis=1)
operational_avg = df[operational_q3].mean(axis=1)

rho_clin_op = clinical_avg.corr(operational_avg, method="spearman")
print(f"Clinical vs operational composite: ρ = {rho_clin_op:.3f}")


Clinical vs operational composite: ρ = 0.591


**Interpretation:**  
*Patients who rate clinical high tend to rate operational high as well (ρ = 0.591) with the ρ being close to 1 than 0*

---

## Section 4 — Visualization

Create at least one visualization that supports one of your analysis findings. Your chart should:

- Have a title that states the finding, not just the data (e.g., "Satisfaction scores drop sharply after age 40" not "Satisfaction by age")
- Have labeled axes
- Use a chart type appropriate for your data (bar for categorical comparison, scatter for relationships, line for trends over time)

Below the chart, explain in a markdown cell: why you chose this chart type, and what you want the reader to take away from it.

In [14]:
sat = "satisfaction in RM"
predictors = [c for c in df.columns if c != sat]

mean_by_factor = df[predictors].mean().sort_values(ascending=False)
plot_df = mean_by_factor.reset_index()
plot_df.columns = ["Factor", "mean_rating"]

fig = px.bar(
    plot_df,
    x="mean_rating",
    y="Factor",
    orientation="h",
    title="Average rating by hospital aspect (which factors score highest vs lowest)",
    labels={
        "mean_rating": "Mean rating (higher = more satisfied with this aspect)",
        "Factor": "Factor",
    },
)
fig.update_yaxes(autorange="reversed")
fig.show()


**Chart rationale:**  
*I wanted to use horizontal bars to make the data easier easier to read, especially because the dataset has many aspect names. While df.describe() can show which category has the highest mean through numeric values on a table, the horizontal bar chart lets a reader compare ranks at a glance and see how far apart the highest and lowest aspects are on the same scale*

In [15]:
# Each dot = one patient: composite score vs satisfaction in RM
# (composite = mean of that group's items for that patient, same columns as the two-bar chart above)

import plotly.express as px

sat = "satisfaction in RM"
personable = [
    "Quality/experience dr.",
    "Admin procedures",
    "Specialists avaliable",
    "friendly health care workers",
    "Communication with dr",
]
operational = [c for c in df.columns if c not in personable and c != sat]

pat = df.copy()
pat["personable_avg"] = df[personable].mean(axis=1)
pat["operational_avg"] = df[operational].mean(axis=1)

pat_long = pd.concat(
    [
        pat.assign(composite_type="Personable", composite_avg=pat["personable_avg"]),
        pat.assign(composite_type="Operational", composite_avg=pat["operational_avg"]),
    ],
    ignore_index=True,
)

fig_both = px.scatter(
    pat_long,
    x="composite_avg",
    y=sat,
    color="composite_type",
    category_orders={"composite_type": ["Personable", "Operational"]},
    opacity=0.35,
    title="Each patient: personable and operational composites vs satisfaction in RM",
    labels={
        "composite_avg": "Mean composite rating (this patient)",
        sat: "Satisfaction in RM",
        "composite_type": "Composite",
    },
)
fig_both.show()

**Chart rationale:**  
*I wanted to do a scatter plot to put two patient-level series on one figure: each person contributes a Personable composite and an Operational composite, both plotted against satisfaction in RM on the y-axis, with color distinguishing the two composites and opacity so dense overlap reads as darker regions. The y-axis shows discrete satisfaction bands (for example stacked near whole-number ratings), while the x-axis is the mean composite rating (1–5) for that patient—so you can compare how the two clouds sit relative to each other without collapsing everything into a single bar of means. That matches the chart’s title and legend (“Composite”: Personable vs Operational) and keeps the unit of analysis at the patient for an association-style read.*

---

## Section 5 — Conclusions

Write 3–5 sentences summarizing what you found. Address these questions:

- What is the most important thing your analysis revealed?
- What surprised you?
- What would you investigate next if you had more time or data?
- What are the limitations of this analysis — what can't you conclude from this data?

**Summary of findings:**  
*The most important thing my analysis revealed was the top factors that were highly associated with the overall satisfaction rating. I was surprised that while exact diagnosis scored higher in rating, communication with doctors was the top factor that was highly associated with the overall satisfaction. If I had more data, I would explore the qualitative data such as the reviews associated with the ratings. I am curious about the context and how qualitative data might support the reasoning behind their rating. The limitations of this analysis is that each rating may be subjective and does not provide enough information, context, or a readme file with the dataset.
Then complete the competency claim below.*

---

## Competency Claim

In a `mp1.md` file in your GitHub repository, write a short competency claim (2–4 sentences) for each domain you feel this project demonstrates. Be specific — cite something you actually did in this notebook.

Domains covered by this project typically include:
- **C3 — Data cleaning and file handling** (if you cleaned or reshaped data)
- **C5 — Data analysis with pandas** (answering questions with code)
- **C6 — Data visualization** (your chart)
- **C7 — Critical evaluation and professional judgment** (your interpretation and limitations section)

You don't have to claim every domain — only the ones your work actually demonstrates.